### PR TITLE
Imran split blue squares permissions

### PR DESCRIPTION
**Pull request by ImzIssa**
# Description
GIves the permission to create or edit or delete blue square to any user.

## Related PRS (if any):
This frontend PR is related to the [PR#943](https://github.com/OneCommunityGlobal/HGNRest/pull/943) PR.

## Main changes explained:
- Added addInfringements, editInfringement and deleteInfringement permission to the lists 
- updated userProfileModal to display correct content

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin user
5. go to dashboard→ Other Links→ Permissions Management→ Manage User Permissions
6. Enter the username of an user that does not have the "Assign" or "Edit"  or "Delete" Blue square permission. Add the permission and Click on Submit.
7.Login as the chosen user (if already logged in, make sure to log out and log back in).
8. Navigate to "User Profile" -> "Blue Squares". Test the permission
9. Visit any other user's profile and test the added permission.
10. repeat steps 6 - 9 for every other blue square permission

## Screenshots or videos of changes:

1. Admin/Owner Permissions Management
Permissions

1. User without "Assign Blue Square" Permission should not be able to see "+" button to add blue square
No Permissions

2. User given "Assign Blue Square" Permission can see see "+" button and can add blue square
Assign Blue

3. User given "Edit Blue Square" Permission should be able to click on a blue square and make changes
Update (2)

4. User given "Delete Blue Square" Permission should be able to click on a blue square and delete that 
Delete
blue square 

5. User given "Edit Blue Square" and "Assign Blue Square" Permission should be able to click on a blue square and make changes and also delete.

Both (2)

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by hbodgal — DISMISSED** [review]
image


everything is working well.

**Comment by nishithaShetty** [comment]
Hi Imran,
Great work!
The functionality works as expected. I tested the add, edit, and delete functionalities both in combination and individually.
PR#2248_0
PR#2248_1
PR#2248_2
PR#2248_3
PR#2248_4
PR#2248_5

**Review by peter6866 — DISMISSED** [review]
Changes work as expected. After assigning the permissions, users previously without the permission are able to assign/delete blue squares.
image
image
image

**Review by cecilia-uu — DISMISSED** [review]
The functionalities of creating, editing and deleting blue squares are perfect.
截屏2024-05-24 下午3 09 02
截屏2024-05-24 下午3 09 53

**Review by Sandhya1236 — DISMISSED** [review]
I successfully followed all the required steps to test the new blue square permissions. After checking out the current branch and running npm install, I cleared the site data/cache and logged in as an admin user. I navigated to Dashboard → Other Links → Permissions Management → Manage User Permissions, and added the "Assign," "Edit," and "Delete" blue square permissions for a user without these permissions. After logging out and back in as the chosen user, I verified that they could access the new permissions under "User Profile" → "Blue Squares" and on other user profiles. This process was repeated for each blue square permission, and all tests were successful, confirming the correct functionality of the permission management feature. Look at the screenshots for your reference.

image
image
image

**Review by Parth-tech — DISMISSED** [review]
Hi Imran,
Everything works well. Upon assigning permission to the user, the user is able to delete, assign/update blue squares. Great job.

Reference for Delete Operation:

*Before Delete Permission:*
Screenshot 2024-05-25 at 1 21 52 AM

*After Delete Permission:*
Screenshot 2024-05-25 at 1 22 57 AM

**Review by Wenboooooo — DISMISSED** [review]
Looks good to me. After assigning the authorization, the user can edit personal blue squares.
image
image

File: src/components/UserProfile/UserProfileModal/UserProfileModal.jsx [hunk 1]
```diff
@@ -376,6 +376,30 @@ const UserProfileModal = props => {
...
+          color="danger"
+          onClick={() => {
+            modifyBlueSquares(id, dateStamp, summary, 'delete');
+          }}
+          style={boxStyle}
+        >
+          Delete
+        </Button>
+        )}
+
```
Comment (nathanah) [inline src/components/UserProfile/UserProfileModal/UserProfileModal.jsx:402]:
```suggestion
```
This is redundant. Can just add permission checks to the 'modBlueSquare' case like so.

```
{type === 'modBlueSquare' && (
          <>
            {canUpdate && 
              <Button
                color="info"
                onClick={() => {
                  modifyBlueSquares(id, dateStamp, summary, 'update');
                }}
                style={boxStyle}
              >
                Update
              </Button>
              }
            {canDelete &&
              <Button
                color="danger"
                onClick={() => {
                  modifyBlueSquares(id, dateStamp, summary, 'delete');
                }}
                style={boxStyle}
              >
                Delete
              </Button>
            }
          </>
        )}
```

File: src/components/UserProfile/UserProfileModal/UserProfileModal.jsx
Comment (nathanah) [inline src/components/UserProfile/UserProfileModal/UserProfileModal.jsx:1]: Just 'viewBlueSquare', 'addBlueSquare', and 'modBlueSquare' are sufficient.

File: src/components/UserProfile/UserProfile.jsx [hunk 2]
```diff
@@ -472,7 +472,7 @@ function UserProfile(props) {
 
     if (type === 'addBlueSquare') {
       setModalTitle('Blue Square');
-    } else if (type === 'viewBlueSquare' || type === 'modBlueSquare') {
+    } else if (type === 'viewBlueSquare' || type === 'modBlueSquare' || type === 'editBlueSquare' || type === 'deleteBlueSquare') {
```
Comment (nathanah) [inline src/components/UserProfile/UserProfile.jsx:475]:
```suggestion
    } else if (type === 'viewBlueSquare' || type === 'modBlueSquare') {
```

File: src/components/UserProfile/BlueSquares/BlueSquare.jsx [hunk 3]
```diff
@@ -24,15 +26,28 @@ const BlueSquare = props => {
...
+                    }else if (canDeleteInfringements){
+                      handleBlueSquare(
+                        canDeleteInfringements,
+                        'deleteBlueSquare',
+                        blueSquare._id,
+                      );
+
+                    }else{
+                      handleBlueSquare(
+                        true,
```
Comment (nathanah) [inline src/components/UserProfile/BlueSquares/BlueSquare.jsx:50]:
```suggestion
```

File: src/components/UserProfile/BlueSquares/BlueSquare.jsx [hunk 4]
```diff
@@ -24,15 +26,28 @@ const BlueSquare = props => {
...
-                    } else if (canPutUserProfileImportantInfo) {
+                    } else if (canEditInfringements && canDeleteInfringements) {
                       handleBlueSquare(
-                        canPutUserProfileImportantInfo,
+                        canEditInfringements && canDeleteInfringements,
                         'modBlueSquare',
                         blueSquare._id,
                       );
-                    } else {
+                    } else if(canEditInfringements) {
```
Comment (nathanah) [inline src/components/UserProfile/BlueSquares/BlueSquare.jsx:35]:
```suggestion
                    } else {
```

File: src/components/UserProfile/BlueSquares/BlueSquare.jsx [hunk 5]
```diff
@@ -24,15 +26,28 @@ const BlueSquare = props => {
                   onClick={() => {
                     if (!blueSquare._id) {
                       handleBlueSquare(isInfringementAuthorizer, 'message', 'none');
-                    } else if (canPutUserProfileImportantInfo) {
+                    } else if (canEditInfringements && canDeleteInfringements) {
```
Comment (nathanah) [inline src/components/UserProfile/BlueSquares/BlueSquare.jsx:29]:
```suggestion
                    } else if (canEditInfringements || canDeleteInfringements) {
```

File: src/components/UserProfile/BlueSquares/BlueSquare.jsx [hunk 6]
```diff
@@ -24,15 +26,28 @@ const BlueSquare = props => {
...
                       handleBlueSquare(
-                        canPutUserProfileImportantInfo,
+                        canEditInfringements && canDeleteInfringements,
```
Comment (nathanah) [inline src/components/UserProfile/BlueSquares/BlueSquare.jsx:31]:
```suggestion
                        canEditInfringements || canDeleteInfringements,
```

**Review by nathanah — CHANGES_REQUESTED** [review]
Looks functionally ok. Remove the redundancies and I'll approve. Might want to create separate API endpoints for each permission/action as I suggested on the backend.

**Review by Infinite-Turkey — DISMISSED** [review]
Left a comment at Back-end side: [Review](https://github.com/OneCommunityGlobal/HGNRest/pull/943#pullrequestreview-2086437054)

**Review by JatinAgrawal94 — DISMISSED** [review]
Feature to grant permission to Add, Delete and Edit a blue square for a user works perfectly fine.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/47699187/b09a7f2d-aca8-4f42-9bc5-6675a539171b

**Comment by ZhangPeizhou** [comment]
Works as expected

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95693786/34ae6c12-b0a9-4a3a-ab97-ac740c5b4d18

**Review by Kavil-Jain-514 — APPROVED** [review]
The functionality works. Great work!!
Before permitting the user, the user cannot add or delete the blue square.
image
image

After giving permission, the user can add or delete the blue squares.
image
image
image

File: src/components/UserProfile/UserProfileModal/UserProfileModal.jsx (hunk 1 above)
Comment (ImzIssa) [reply src/components/UserProfile/UserProfileModal/UserProfileModal.jsx]: I've made the changes. Please review again

**Review by mohabbasd — DISMISSED** [review]
Hello, I tested your PR and it works as expected. Great Job! Below is a video for your reference.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91215370/d32b57e6-ba2f-406f-8d50-9a490a05657c

**Review by CloudyZ524 — COMMENTED** [review]
Hi Imran,
The functionality of adding blue square permission "Assign", "Edit", "Delete" to a user and edit the blue square to a user works well from my end. 

Screenshot 2024-06-06 at 13 47 37

However, I encountered an issue that when I add a new blue square, I got the successful and failed message at the same time. 

Screenshot 2024-06-06 at 13 55 03

**Comment by akshit0211** [comment]
The functionality of assign, edit, and delete works as expected. Good job!

permission

Without permissions.
without permission

assign
assign

edit
edit

delete
delete

**Review by akshit0211 — APPROVED** [review]

**Review by Sudershhh — DISMISSED** [review]
Created accounts for testing this PR. 

I am able to assign, edit and delete blue squares. 

PR 2248

PR 2248_2

**Review by PabloWang — APPROVED** [review]
Hi Imran,
Good job! All changes work as expected. I tested thre add, edit, and delete the blue squares after assigning the permissions to the user.

Add a blue squares after assigning the permissions:
Screenshot 2024-06-16 at 12 35 10 AM

Screenshot 2024-06-16 at 12 36 31 AM

**Review by shradac — DISMISSED** [review]
Looks good to me. After assigning the authorization, the user can add, update, delete personal blue squares.

Screenshot 2024-06-20 at 12 18 57 AM
Screenshot 2024-06-20 at 12 41 43 AM

**Review by DiyaWadhwani — CHANGES_REQUESTED** [review]
Hey Imran, tested this out as per your instructions.
I logged into an owner account and gave the Assign, Edit and Delete Blue squares permissions to the volunteer role.

Logging into the volunteer account gives the volunteer user these accesses. However, when I logged back into the owner account and deleted the Edit Blue Squares permission, ideally the volunteer should not be seeing an 'Update' button on either of the blue squares. Similarly for when the Delete Blue Squares permission is revoked and in the case when the Assign blue squares permission is revoked, the user should not be seeing the '+' button. 
I have tested this on the chrome browser and then tested it again on an incognito window and reloaded the window before every login. I have attached a video for your reference.
Kindly review this.
Thank you.

Uploading 943+2248.mp4…

**Review by vishavk1992 — DISMISSED** [review]
I have tested this PR alongside backend PR 943, and it meets the requirements. After assigning permissions from my owner account to the volunteer account, the "+" sign for the blue square appeared. Upon clicking it, I could update and delete the blue square. After revoking the permission, assigning and updating the blue square successfully disabled it from the volunteer account.  attached screenshots

PR-2248+943
PR-2248
pr-2248-1
pr-2248-2

**Review by nikhilpittala16 — APPROVED** [review]
I have tested the PR and everything works as expected.
Screenshot (37)
Screenshot (38)
Screenshot (40)
Screenshot (41)
Screenshot (42)
Screenshot (43)

**Review by Vigneshwar-Muriki — DISMISSED** [review]
Tested this PR and it works as expected for all three scenarios.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/131609223/4aa4f3c9-87ce-4310-917f-b62823541c92


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/131609223/9aea0c18-893a-4d37-bd59-dad9e85b8abc


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/131609223/02ebeee8-a784-463c-9e4b-0306452dc1a0

**Review by nyelmela — DISMISSED** [review]
I have tested this PR, and everything works perfectly well. I checked into the current branch, ran `npm install`, and executed the necessary commands to run the PR locally. After clearing the site data and cache, I logged in as an admin user and navigated to the dashboard → Other Links → Permissions Management → Manage User Permissions. I entered the username of a user who did not have the "Assign", "Edit", or "Delete" Blue square permission, added the permission, and clicked on Submit. I then logged in as the chosen user (making sure to log out and log back in if already logged in), navigated to "User Profile" → "Blue Squares", and tested the permission. Additionally, I visited another user's profile to test the added permission. I repeated steps 6-9 for every other Blue square permission, and everything worked as expected.

image
image
image
image

**Comment by angelalalacheng** [comment]
Hi, 
I use your frontend branch `Imran_Split_Blue_Squares_Permissions` and backend branch `Imran_Split_Blue_Squares_Permissions`.  But, for me, it seems like some errors even if I give the permission to volunteer account.

Screenshot 2024-07-12 at 8 45 58 PM
Screenshot 2024-07-12 at 8 59 32 PM
Screenshot 2024-07-12 at 8 59 56 PM
Screenshot 2024-07-12 at 9 12 47 PM

**Review by AnujVakil — APPROVED** [review]
I have tested this PR and working as expected.
PR #2248_1
PR #2248_2

**Review by SaiLavanya1 — DISMISSED** [review]
I've tested the PR #2248,The permissions for creating, editing, and deleting Blue Squares were successfully assigned to the specified user.All added permissions were verified and are functioning correctly without any issues.
Screenshot 2024-07-20 at 10 14 41 AM
Screenshot 2024-07-20 at 10 17 28 AM
Screenshot 2024-07-20 at 10 25 02 AM
Screenshot 2024-07-20 at 10 29 57 AM

**Review by ishan-miglani — APPROVED** [review]
Hi Imran, I have reviewed the changes in this PR. After assigning the permissions to a volunteer account from an owner account, I am able to add, edit and delete blue squares. On removing permissions one by one the functionalities are also restricted correctly. Thanks!

https://github.com/user-attachments/assets/b222752d-f162-4e73-9a7f-ec25eb2565f2

**Review by Pranay-ai — APPROVED** [review]
Hey @ImzIssa 
I have tested the PR, assigned permission to user, and the user is able to delete and assign blue squares
image
image
image

Thank You..

**Review by adityasure9 — APPROVED** [review]
Hi Imran,

The permissions for adding, editing, and deleting Blue Squares work as expected. I tested the feature thoroughly, and everything functions correctly. Great job on this!

**Comment by Lemonnycodes** [comment]
Hi,

- Verified the functionality of the "Assign," "Edit," and "Delete" Blue Square permissions:
- Added each permission to a user who initially did not have them.
- Logged in as that user to confirm that the permissions were correctly applied.
- Tested each permission by navigating to the "User Profile" → "Blue Squares" section and visiting other users' profiles.
- Confirmed that the user could perform the actions associated with each permission as expected.
- Repeated the process for all three Blue Square permissions to ensure they worked correctly.
- The PR works as intended, and the permissions are applied and tested successfully
Screenshot 2024-08-14 011037
Screenshot 2024-08-14 011058
Screenshot 2024-08-14 011218

**Review by SMitta11 — APPROVED** [review]
The functionalities given in this PR is working as expected.
I gave volunteer user edit, delete and add permission for blue square from admin account and tested these individually and in combination. These are working fine.
Screenshots:
Screen Shot 2024-08-17 at 12 03 13 AM
Screen Shot 2024-08-17 at 12 03 46 AM
Screen Shot 2024-08-17 at 12 04 48 AM
Screen Shot 2024-08-17 at 12 05 10 AM

**Review by Lemonnycodes — DISMISSED** [review]
Hi,
The PR works as mentioned in the description, and the permissions are applied and tested successfully
- Verified the functionality of the "Assign," "Edit," and "Delete" Blue Square permissions:
- Added each permission to a user who initially did not have them.
- Logged in as that user to confirm that the permissions were correctly applied.
- Tested each permission by navigating to the "User Profile" → "Blue Squares" section and visiting other users' profiles.
- Confirmed that the user could perform the actions associated with each permission as expected.
- Repeated the process for all three Blue Square permissions to ensure they worked correctly.

Screenshot 2024-08-14 011037
Screenshot 2024-08-14 011058

**Review by nathanah — APPROVED** [review]

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**